cmd/s3-layer.go

```diff
@@ -0,0 +1,508 @@
+	default:
+		fmt.Println("Undefined s3 object error:", bucket, object, minioErr.Code)
+	}
+
+	e.e = err
+	return e
+}
+
+type s3LayerConfig struct {
+	Location        string
+	Endpoint        string
+	AccessKeyID     string
+	SecretAccessKey string
+	UseSSL          bool
+}
+
+// S3Layer - Implements Object layer for S3 and Minio blob storage.
+type S3Layer struct {
+	s3LayerConfig
+
+	Client *minio.Core
+}
+
+func newS3Layer(endpoint, location, accessKey, secretKey string) (GatewayLayer, error) {
+	config := s3LayerConfig{
+		Endpoint:        endpoint,
+		Location:        location,
+		AccessKeyID:     accessKey,
+		SecretAccessKey: secretKey,
+		UseSSL:          true,
+	}
+
+	// Initialize minio client object.
+	client, err := minio.NewCore(config.Endpoint, config.AccessKeyID, config.SecretAccessKey, config.UseSSL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &S3Layer{
+		s3LayerConfig: config,
+		Client:        client,
+	}, nil
+}
+
+// Shutdown - save any gateway metadata to disk
+// if necessary and reload upon next restart.
+func (l *S3Layer) Shutdown() error {
+	// TODO
+	return nil
+}
+
+// StorageInfo - Not relevant to S3 backend.
+func (l *S3Layer) StorageInfo() StorageInfo {
+	return StorageInfo{}
+}
+
+// MakeBucket - Create a new container on S3 backend.
+func (l *S3Layer) MakeBucket(bucket string) error {
+	err := l.Client.MakeBucket(bucket, l.Location)
+	if err != nil {
+		return s3ToObjectError(traceError(err), bucket)
+	}
+	return err
+}
+
+// GetBucketInfo - Get bucket metadata..
+func (l *S3Layer) GetBucketInfo(bucket string) (BucketInfo, error) {
+	buckets, err := l.Client.ListBuckets()
+	if err != nil {
+		return BucketInfo{}, s3ToObjectError(traceError(err), bucket)
+	}
+
+	for _, bi := range buckets {
+		if bi.Name != bucket {
+			continue
+		}
+
+		return BucketInfo{
+			Name:    bi.Name,
+			Created: bi.CreationDate,
+		}, nil
+	}
+
+	return BucketInfo{}, traceError(BucketNotFound{Bucket: bucket})
+}
+
+// ListBuckets - Lists all S3 buckets
+func (l *S3Layer) ListBuckets() ([]BucketInfo, error) {
+	buckets, err := l.Client.ListBuckets()
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]BucketInfo, len(buckets))
+	for i, bi := range buckets {
+		b[i] = BucketInfo{
+			Name:    bi.Name,
+			Created: bi.CreationDate,
+		}
+	}
+
+	return b, err
+}
+
+// DeleteBucket - delete a bucket on S3
+func (l *S3Layer) DeleteBucket(bucket string) error {
+	err := l.Client.RemoveBucket(bucket)
+	if err != nil {
+		return s3ToObjectError(traceError(err), bucket)
+	}
+	return nil
+}
+
+// ListObjects - lists all blobs in S3 bucket filtered by prefix
+func (l *S3Layer) ListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+
+	loi := ListObjectsInfo{}
+
+	loi.Objects = make([]ObjectInfo, maxKeys)
+
+	i := 0
+
+	for message := range l.Client.ListObjectsV2(bucket, prefix, true, doneCh) {
```

Comment:
you can use https://github.com/minio/minio-go/blob/master/core.go : ListObjects()